Source/CarthageKit/DiagnosticResolver.swift

```diff
@@ -0,0 +1,170 @@
+
+/**
+Delegate for listening to resolver events.
+*/
+public protocol DiagnosticResolverDelegate: class {
+	func diagnosticResolver(_ resolver: DiagnosticResolver, didFindVersions versions: [PinnedVersion], compatibleWith versionSpecifier: VersionSpecifier, for dependency: Dependency)
+	func diagnosticResolver(_ resolver: DiagnosticResolver, didFindTransitiveDependencies transitiveDependencies: [(Dependency, VersionSpecifier)], for dependency: Dependency, withVersion version: PinnedVersion)
+	func diagnosticResolver(_ resolver: DiagnosticResolver, didEncounterError error: Error, whileRetrievingTransitiveDependenciesFor dependency: Dependency, withVersion version: PinnedVersion)
+	func diagnosticResolver(_ resolver: DiagnosticResolver, didEncounterError error: Error, whileRetrievingVersionsFor dependency: Dependency, compatibleWith versionSpecifier: VersionSpecifier)
+}
+
+/**
+Resolver which logs all dependencies it encounters, and optionally stores them in the specified local repository.
+*/
+public final class DiagnosticResolver: ResolverProtocol {
+	public weak var delegate: DiagnosticResolverDelegate?
+	public var localRepository: LocalRepository?
+	public var ignoreErrors: Bool = false
+	private let versionsForDependency: (Dependency) -> SignalProducer<PinnedVersion, CarthageError>
+	private let resolvedGitReference: (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>
+	private let dependenciesForDependency: (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>
+
+	private var versionSets = [Dependency: [PinnedVersion]]()
+	private var handledDependencies = Set<PinnedDependency>()
+
+	// Specify mappings to anonimize private dependencies (which may not be disclosed as part of the diagnostics)
+	public var dependencyMappings: [Dependency: Dependency]?
+
+	public init(
+		versionsForDependency: @escaping (Dependency) -> SignalProducer<PinnedVersion, CarthageError>,
+		dependenciesForDependency: @escaping (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>,
+		resolvedGitReference: @escaping (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>
+		) {
+		self.versionsForDependency = versionsForDependency
+		self.dependenciesForDependency = dependenciesForDependency
+		self.resolvedGitReference = resolvedGitReference
+	}
+
+	public func resolve(
+		dependencies: [Dependency: VersionSpecifier],
+		lastResolved: [Dependency: PinnedVersion]? = nil,
+		dependenciesToUpdate: [String]? = nil
+		) -> SignalProducer<[Dependency: PinnedVersion], CarthageError> {
+		let result: Result<[Dependency: PinnedVersion], CarthageError>
+
+		do {
+			try traverse(dependencies: Array(dependencies))
+			result = .success([Dependency: PinnedVersion]())
+		} catch let error {
+			let carthageError: CarthageError = (error as? CarthageError) ?? CarthageError.internalError(description: error.localizedDescription)
+			result = .failure(carthageError)
+		}
```

Comment:
[nits] I think it's better to split error handling by types:
```
do {
    try traverse(dependencies: Array(dependencies))
    result = .success([Dependency: PinnedVersion]())
} catch let error as CarthageError {
    result = .failure(error)
} catch {
    result = .failure(CarthageError.internalError(description: error.localizedDescription))
}

Reply:
Applied this change